### PR TITLE
Infer source prefix automatically

### PR DESCRIPTION
**Pull request by avdv**
- **Determine source prefix automatically**
- **Use inferred source_prefixes in _compile_module**
- **Move package conf actions into dynamic ouput**
- **Infer source prefixes for exposed-modules and import-dirs**
- **Properly handle html output with source prefixes**
- **Fix handling of output directories**
- **Handle source prefix for boot modules**
- ... revert changes using the `src_strip_prefix` attribute

### CODE REVIEW

**Review by aherrmann — COMMENTED** [review]
Thank you! That looks good. One comment on the `-i` flags though.

File: haskell/compile.bzl [hunk 1]
```diff
@@ -599,6 +584,20 @@ def _compile_module(
         children = [cross_package_modules] + this_package_modules,
     )
 
+    # add each module dir prefix to search path
+    source_prefixes = dependency_modules.reduce("source_prefixes").keys()
```
[the remark this answers is not in the document]
Comment (avdv) [reply haskell/compile.bzl]: Oh, yes of course! Thank you. I removed the reduction function and the tracking of the source prefixes in the `CompiledModuleInfo` again.

**Comment by avdv** [comment]
Please note, this change still does not support multiple source directories per package since currently we require that the haskell_metadata action has access to the source files of cross package modules.

I'll work on this next, but this PR could be merged already as it contains a fix for the problem with `external_tools` with `haskell_binary` rules.

**Review by aherrmann — APPROVED** [review]
Looks good, thank you! One small comment.

File: haskell/haskell_haddock.bzl [hunk 2]
```diff
@@ -65,9 +69,19 @@ def _haddock_dump_interface(
         for dep_name in graph[module_name]
     ]
 
+    expected_html = outputs[haddock_info.html]
+    module_html = _haddock_module_to_html(module_name)
+
+    if paths.basename(expected_html.short_path) != module_html:
+        html_output = ctx.actions.declare_output("haddock-html", module_html)
+        make_copy = True
```
[the remark this answers is not in the document]
Comment (avdv) [reply haskell/haskell_haddock.bzl:77]: I thought so too and tried that before, but that did not work. The error I got was about the output artifact that was expected to be bound already:
```
BUILD FAILED
for action key `(target: `root//backend:backend_infra (prelude//platforms:default#904931f735703749)`, id: `1522`)`

Caused by:
    Traceback (most recent call last):
      File <builtin>, in <module>
      * prelude/haskell/haskell_haddock.bzl:177, in haddock_dump_interfaces
          module_tsets[module_name] = _haddock_dump_interface(
      * prelude/haskell/haskell_haddock.bzl:78, in _haddock_dump_interface
          ctx.actions.copy_file(expected_html.as_output(), html_output)
    error: artifact `(root//backend:backend_infra (prelude//platforms:default#904931f735703749))/haddock-html/Model-Numeric.html` must be bound by now. If you are intending to use this artifact as the output of `run`, are you missing an `.as_output()` call?
      --> prelude/haskell/haskell_haddock.bzl:78:9
       |
    78 |         ctx.actions.copy_file(expected_html.as_output(), html_output)
       |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       |
```

[the remark this answers is not in the document]
Comment (aherrmann) [reply haskell/haskell_haddock.bzl:77]: Interesting, TIL!

**Pull request merged**